### ISSUE TITLE
[Suggestion] Package Manager Manager GUI based on Pamac to manage Bedrock Linux

### ISSUE THREAD

**Issue by setsunati** [bedrocklinux/bedrocklinux-userland#100]
Visual representation:  https://i.imgur.com/b5kzrhH.png

**Comment by paradigm** [comment]
It'll be a while before we get to writing pmm.  Once we do, we probably won't *automatically* distribute a GUI for it, as Bedrock strives to have its core be relatively small.  However, we could offer an optional GUI wrapper for it.  May be a bit, as I'm inclined to prioritize other things over it, but if no one else does it before I get there, I may write such a thing eventually.

I'm not familiar with Pamac - is it configurable to various back-end package managers?  If so, and we could just distribute a configuration file for it, that'd be perfect.  I don't mind distributing a small configuration file that works with a GUI wrapper other distros distribute.

**Comment by setsunati** [comment]
Don't know the answer to that. Hopefully someone else can answer your question.

**Comment by xacrimon** [comment]
I'd be happy to contribute, I would probably be interested in writing a full combined package package manager with support for all strata if, that's desired. I myself have a lot of experience in python, go and rust. Looking to start contributing heavily soon @paradigm

**Comment by xacrimon** [comment]
Initially I would probably write a cli backend and Frontend and gui later if desired.

**Comment by xacrimon** [comment]
What do you think?

**Comment by paradigm** [comment]
I've put a *lot* of thought into `pmm`'s architecture and have a lot of it mentally sketched, but I think there's still more research before you or me or anyone else should start coding its CLI.  Some of the problems are kind of tricky.

However, I've not put much thought into a GUI at all as I don't personally have much interest in it, and handing that off to someone else is probably a good idea.  If you're itching to start working and have an interest here, preliminary work on a GUI layer that we can slot the CLI into later seems reasonable to me.

Various disorganized thoughts:

- While we get a non-trivial amount of drive-by support, we're very limited in terms of long term maintenance manpower.  If you lose interest and disappear, we want to ensure someone else can maintain it, and we don't have many GUI people.  Try to keep it simple and clean and small.
- Bedrock's code base is entirely statically linked C99 and busybox shell.  I don't want to introduce new runtime dependencies, which I think would rule out python.  I'd rather minimize build-time dependencies as well, if possible.  I'd also like to avoid requiring other Bedrock devs learn new languages to maintain it.  If C99/shell is feasible for you, I prefer we stick to it.  However, if you have strong objections to it, I can probably be talked into Golang or Rust.  If you know both Golang and Rust, I expect C should come quickly to you.
- People use Bedrock to use features from other distros, not Bedrock itself.  Bedrock is, in some sense, pure overhead.  Thus, we strive to keep Bedrock small.  Whatever strategy is used here (e.g. GUI toolkits), aim for minimizing disk usage (within reason).
- Adding new third party libraries into Bedrock's build system, e.g. a GUI toolkit, is okay so long as we keep it small.  However, there's also the possibility of detecting and using some standard GUI toolkit from other strata.  Since this an optional feature, I don't mind saying users have to install GTK or QT or something in another stratum for it to work.  If you consider going down that road, test against a bunch of different distros implementations of that library - all the glibc ones in `brl fetch --list` and `brl fetch --experimental`, for example, to ensure portability, as that's kind of essential for Bedrock.  If you can make it work with both glibc and musl ones that'd be great, but I don't know the feasibility of such things.
- [This](https://bedrocklinux.org/future.html#pmm) is the entirety of pmm's current docs.  There's a lot in my head I didn't yet write down.  Hopefully that's enough to get started.  If not, I'll see if I can pull pmm up on my priority list and get more down to share sooner.
- You could also start with a GUI layer over `brl`.  For example, a GUI that can list strata/aliases and their statuses, or help fetch new strata.  Could be a good initial target to get a feel for language, toolkit, etc options while `pmm` research is still under way.

Lemme know if that sounds reasonable and interesting.

**Comment by xacrimon** [comment]
Okay, I'm not as comfortable in C as Rust but 7 could make do. C99 sounds pretty reasonable as well but I am 10x better in Rust. I would much rather do it in rust. With some clever tactics we can keep disk low aswell. I could start working on a GUI but I'd like to atleast implement a C/Rust backens library first

**Comment by xacrimon** [comment]
Could make made to work with both by only statically linking musl

**Comment by xacrimon** [comment]
But compiling should work for both no problem.

**Comment by paradigm** [comment]
> Okay, I'm not as comfortable in C as Rust but i could make do. C99 sounds pretty reasonable as well but I am 10x better in Rust. I would much rather do it in rust.

If you can make due with C99, that may be best.  Minimizes runtime dependencies, build dependencies, and background needed to maintain Bedrock, as well as improves portability (e.g. if we want to do a RISC-V port).  I think all the performance sensitive bits will be in the CLI tool or a GUI library such that expert knowledge in the language used for the GUI layer probably isn't necessary.  For this, we'd only require adequate knowledge to write something clean and stable.  Plus, given the fact some of it is blocked by yet to be completed research and design work, there's time to improve one's C.

> With some clever tactics we can keep disk low aswell.

Excellent.

> I could start working on a GUI but I'd like to atleast implement a C/Rust backens library first

My primary concern here is that the specifics for the back-end design are still very much in flux.  I have a number of possible designs in mind with different trade offs that need to be settled, as well as open design problems that need to be addressed.

If you want to make something as a placeholder back end to help work on a GUI I certainly have no objection, so long you're comfortable with the whole thing getting tossed out.

While it's possible it'll change later, I'm currently leaning towards a pure shell implementation of `pmm`, which would mean a C99 GUI layer can't link to it.  If you want to make a placeholder back-end, consider having a clean separation with communication over `fork()`/`exec()` and pipes.

> Could make made to work with both by only statically linking musl

If you can make an executable that's statically linked against musl but will dynamically link against glibc toolkits from other strata, that would be great!  I wasn't sure if such things were possible.  Maybe there's some linker magic with which I'm unfamiliar, or I'm imagining non-existent constraints.

**Comment by xacrimon** [comment]
I don't really see the problem with rust as it supports most architectures and has no runtime dependencies. What I would like to do is start working in a personal repository and then if and when the project becomes very useful, merging into mainline bedrock. Rust has Tier 2 support, all this means is that the compiler is not available for Arm71. Its still maintained etc. If desired, you could  start work on a shell version of pmm. Regardless I'll be writing a pmm for myself which we could consider later down the line.

**Comment by xacrimon** [comment]
You could look at redox which handles this very well imo

**Comment by paradigm** [comment]
> I don't really see the problem with rust

- Increases build-time requirements non-trivially.
- Increases background needed to maintain Bedrock.  If you lose interest and leave, I have to consider spending a lot of time to learn Rust at the expense of other Bedrock efforts (which is going to upset a lot of Bedrock users, including myself) or drop the feature (which is going to upset a lot of Bedrock users, which in turn is going to upset me).
- Second-tier support for an architecture Bedrock is currently supporting (armv7l).
- *Zero* support for other architectures Bedrock can currently be self-compiled to (microblaze), removing the option of supporting it down the line.

These aren't completely insurmountable problems.  If there was an adequately strong benefit to Rust, we could look past the problems.  However, the case for it doesn't seem very strong here.

> What I would like to do is start working in a personal repository and then if and when the project becomes very useful, merging into mainline bedrock.

Odds are very slim that it will be merged into mainline Bedrock if you develop it independently.  I'm not trying to be mean or stubborn here.  Bedrock is very strapped for manpower and I don't want to turn away a helping hand.  However, if you're planning on just ignoring all of the thought, discussions, research, prototypes, etc odds are whatever you come up with is going to end up fairly distant from what the community is currently expecting.

You're absolutely welcome to make something for your own personal use.  Other people have made automated Bedrock installers, package managers wrappers, and other features for their own use to tide them over until Bedrock's official one comes around.  All I ask there is that you name it something other than "pmm" to avoid confusion with Bedrock's official efforts here.

**Comment by xacrimon** [comment]
I'll reconsider C99, if I am going to contribute to bedrocks official pmm, do note that I'm quite uncomfortable with creating GUIs in C. I would really like to help bedrock even if it means getting really good at C. If you me to help, make that clear. I'll brush up my C99 if that's the case. If you do not currently want contributions regarding pmm I would be okay with waiting and building my own in the meantime.

**Comment by SolitudeSF** [comment]
i dont think gui should be a part of bedrock/pmm, it should be just a wrapper around cli/library and distributed separately.

**Comment by Mouvedia** [comment]
IMHO @xacrimon you should join \#bedrock and talk with paradigm directly.

Secondly we should list the major package managers that pmm will target:

- DNF/YUM
- eopkg/PiSi
- apt/dpkg
- Nix
- Portage
- Entropy
- pacman
- Conary
- [etc.](http://www.self.gutenberg.org/articles/List_of_software_package_management_systems)

**Comment by xacrimon** [comment]
We definently should

**Comment by Mouvedia** [comment]
Also @paradigm you should take a look at https://labix.org/smart

**Comment by ryleu** [comment]
I'm very interested in the development here. I'm working on some basic Rust UI stuff, and I'm planning on just keeping a basic list of what distros have which package managers for now, at least until PMM gets some sort of universal CLI backend.

I don't think writing GUI stuff in C99 is a good idea, at least for my sanity. This project wouldn't be a part of the main bedrock install, and I would instead prefer if it was a fully static binary, so it doesn't interfere with any of the base system's functionality.

I would also probably have to write it for GTK, LibAdwaita, *and* Qt5, so it looks good in all contexts, which sounds like a massive pain. I could just have three binaries on the releases page and a few build flags set up for different systems.

My current work is purely learning, so I don't have a git repo yet, but I'll keep y'all updated once I do. The plan for now is just to get a UI working with a temporary wrapper library for a fake PMM of my own making that can later be swapped for the real deal..

**Comment by paradigm** [comment]
A lot of the blockers from the last time we discussed this in January 2019 have been resolved over the intervening years such that it now seems feasible for @ryleu or @xacrimon work on it:

- Naga's code base is planned to support Rust for some components. [^1]
- Naga's `pmm` is planned to be one of the items written in Rust. [^2]
- Naga _may_ include its own (relatively simple) package manager. [^3] [^4]

I'm disinclined from making Naga's code public until it's ready for a public alpha [^5].  However, Naga's (cli, written in Rust) `pmm` is likely to be something that can be initially developed against Poki without much issue.  It's very much something I can delegate to others who could start work on it whenever, if someone is interested. [^6]  This is probably a prerequisite for a hypothetical graphical `pmm`, but once Naga's cli `pmm` is far enough along I see no reason to avoid or delay work on a graphical, written in Rust `pmm`.

[^1]: In January 2019, Bedrock was mostly carefully curated system calls with little threading or typed business logic, which greatly limits Rust's strengths over C.  Moreover, the projected architecture support roadmap hit a lot of Rust Tier 3 or not even Tier 3 items.  However, the future roadmap involves both more subtle threading and typed business logic where Rust's case is much stronger than C's.  That having been said, in my experimentation with Rust it _still_ falls flat in many cases; Naga's initial release will likely be a mix of Rust and C.  (I'll be game for someone presumably more Rust savvy to port the initial C versions of things to Rust after 0.8.0, if someone can do so without the CPU/RAM/disk hit I couldn't find a way to avoid.)  The fact that _zero_ of Bedrock's target architectures are Rust Tier 1 makes me a bit nervous here, and we'll probably have to drop support for some of the more obscure currently supported ISAs (e.g. i486), but given the threading and logic complexity of some Naga components I think the trade-off will be worthwhile.

[^2]: This means a GUI won't have to _just_ wrap `pmm`, but could share much of the logic.

[^3]: If the default install ends up lacking a GUI package manager, users who are interested in it could still trivially install it via the package manager.  Hypothetically if there's some unfortunate point of disagreement and the GUI doesn't end up being officially part of Bedrock, it could still be in an unofficial third party repo the package manager would cleanly support.

[^4]: My main impetus here is to disconnect Bedrock's version from the `brl fetch` scripts that keep breaking as distros change; having Bedrock's core be a separate package makes Bedrock's core change log must less noisy.  This would also be of value to minimalists as they can then remove non-essential things like (the cli) `pmm` if they don't use it, or even `brl fetch` once they've got their desired set of strata.

[^5]: In my experience, doing so just results in people asking questions due to idle curiosity and eating what otherwise would have been productive time.

[^6]: If someone is interested, let me know and I can spell out the plans and explain how it will differ from Poki's version.

**Comment by ryleu** [comment]
> The fact that zero of Bedrock's target architectures are Rust Tier 1 makes me a bit nervous here

Tier 1 Rust includes x86_64, i686, and aarch64. What do you mean by "target architectures"?

> Naga's `pmm` is planned to be one of the items written in Rust.

That's good to hear, wrapping is nowhere near as clean as an API / logic sharing would be.

> If the default install ends up lacking a GUI package manager, users who are interested in it could still trivially install it via the package manager.

I actually *prefer* the idea of having it be optionally available through Naga's package manager, for which I would like to suggest the name `bpm` (Bedrock Package Manager).

> Hypothetically if there's some unfortunate point of disagreement and the GUI doesn't end up being officially part of Bedrock, it could still be in an unofficial third party repo the package manager would cleanly support.

I can see a lot of possible causes for disagreement, especially considering the many different setups (GTK, Qt, LibAdwaita, and whatever else) that I'd need to support. My original plan was to just distribute it as a tarball, but I prefer the idea of having it available in `bpm`.

I could start work on converting `pmm` to Rust, but it would take a lot of time because I'm not particularly familiar with how it works.

---

This part really should be another comment, but I don't want to send two emails.

My plan right now is to have GUI `pmm` work similarly, at least UX-wise, to `gnome-software`, specifically in the way that it manages packages from different sources:

![image](https://user-images.githubusercontent.com/69326171/164910346-7d4222c9-e797-4b29-93a8-36d8fd8d1cb3.png)

**Comment by paradigm** [comment]
> > The fact that zero of Bedrock's target architectures are Rust Tier 1 makes me a bit nervous here
> 
> Tier 1 Rust includes x86_64, i686, and aarch64. What do you mean by "target architectures"?

x86_64-unknown-linux-gnu, i686-unknown-linux-gnu, and aarch64-unknown-linux-gnu are Tier 1 Rust.  x86_64-unknown-linux-musl, i686-unknown-linux-musl, and aarch64-unknown-linux-musl are Tier 2.

glibc is problematic for Bedrock's use case:

- It is not static linking friendly. [^1]  Certain Bedrock strictly need to be statically linked due to the fact they need to run in different environments.
- As Bedrock approaches 1.0, sensitivity to things like how much disk space it takes up become a higher priority. [^2] musl is much more size-conscious than glibc. [^3]


> I actually _prefer_ the idea of having it be optionally available through Naga's package manager, for which I would like to suggest the name `bpm` (Bedrock Package Manager).

I'm sensitive to concerns over clobbering someone else's name. [^4] I also like bpm, but [someone already has a `bpm` binary in major distros](https://packages.debian.org/buster/amd64/bpm-tools/filelist).  My choice would have been bpkg, but that's used used [by](http://freshmeat.sourceforge.net/projects/bpkg) multiple [people](https://github.com/bpkg/bpkg) already.  At least one of them has a `bpkg` binary in Gentoo.

I'm currently tentatively using Bedrock Package Tool, `bpt`, borrowing from Debian.  Another option on the table is `bpk`, borrowing from Alpine.

> I could start work on converting `pmm` to Rust

To make sure there's no confusion by anyone reading along, the goal here isn't to rewrite something in Rust for its own stake.  Rather, Naga's `pmm` is planned to be quite different from Poki's such that we need a rewrite, and since we're rewriting anyways Rust makes sense here.

Sadly I've not been very organized about collecting thoughts here [^5], but off the top of my head some of the changes for Naga are:

- Performance improvements.  In particular, Poki `pmm`'s logic to scan which package managers are available is excruciatingly slow; so much so that it's cached, but sadly that [introduces the famously finicky problem of cache invalidation](https://twitter.com/secretGeek/status/7269997868).  If available, Rust (or C) could use `openat2` or iu-uring to do this much more quickly, and only fall back to Poki `pmm`'s strategy for older kernels.
- The ability to run package manager operations in parallel.  The main thing that makes this tricky is the fact package managers are often interactive, e.g. confirmation prompts.  I'd like to resolve this by having `pmm` forward stdin/stdout/stderr for one of them at a time, then pop up the next still-running one once the first one is done.  To make the situation clear to the end user, I'd like to reserve a line at the bottom of the terminal that prints `pmm`'s parallel operation status, [similar to apt](https://149366088.v2.pressablecdn.com/wp-content/uploads/2014/04/apt-progress.png).  If `apt` runs in `pmm`, it should nest; you'd have two reserved lines at the bottom.
- Support for per-user configuration, rather than (just) system wide.  One user should be able to set the `pmm` user interface differently from another.
- As part of a broader Naga change, system-wide settings should be in `</bedrock/strata/<pmm-stratum>/etc` rather than `/bedrock/etc`.  In Naga, Bedrock-specific things will be in the `bedrock` stratum (e.g. `/bedrock/strata/bedrock/etc/pmm.conf` or whatever); `/bedrock` outside of `/bedrock/strata` will only be about cross-stratum stuff.
- The ability to set the interface on-the-fly, e.g. via an environment variable or flag.  This will make it easier to script around.
- While it's a low priority, eventually [the ability to run on a traditional distro, not just Bedrock](https://github.com/bedrocklinux/bedrocklinux-userland/issues/261).  If this is kept in mind from the beginning, it should be easy to support, I think.  If `pmm` detects it's not on Bedrock, just detect local package managers instead of searching cross stratum and drop the `strat <stratum>` bits.  The other logic should be largely the same.
- Poki `pmm`'s syntax to configure package manager support was specifically chose to make it easy to support in `awk`.  This obviously provides no value for a Rust implementation and will likely change for Naga.  I haven't put thoughts into specifics here, though.  Other Naga components will probably be [toml](https://crates.io/crates/toml); that should be considered here.

If you're serious about doing this make sure I know so we don't end up with multiple people independently working on incompatible efforts.

> but it would take a lot of time because I'm not particularly familiar with how it works.

If you're a reasonably experienced developer, I don't think you'll have much trouble figuring out the broad strokes of Poki's `pmm`.  I comment liberally if I'm concerned something isn't obvious in context [^6], and most of the business logic parts is pretty straight forward.  The ugliest parts are:

1. String escaping which results from the fact it's written in nested stringly typed languages.  A Rust rewrite won't have to deal with that.
2. Hacking round the fact the language(s) it's written are pretty limited in terms of data structure availability.  Rust shines here.

I think the difficult parts will be the new stuff I listed above rather than porting any of the existing stuff.  If you aren't intimately familiar with vt100 escape codes, for example, it may take a bit to figure out how to reserve a line at the bottom of the terminal.

I'm sympathetic to the idea that one wouldn't _want_ to take a lot of time to work on something like this; I don't want to you to feel like I'm putting pressure on you here.  However, if that it taking you a while to learn how Poki's `pmm` works is your only concern, do note that it'll likely be faster for you to figure out Poki's `pmm` code base than for me to go through all the other Naga stuff I have to do before I can start on `pmm` myself.

> This part really should be another comment, but I don't want to send two emails.
> 
> My plan right now is to have GUI `pmm` work similarly, at least UX-wise, to `gnome-software`, specifically in the way that it manages packages from different sources:
> 
> ![image](https://user-images.githubusercontent.com/69326171/164910346-7d4222c9-e797-4b29-93a8-36d8fd8d1cb3.png)

I, personally, have zero thoughts on how a GUI `pmm` should work, and so I likely won't respond to things like this.  I want to be clear that it's not because I'm dismissive or discouraging of the idea, but just that I don't have anything to say on it.  Do feel free to explore the area here with others who are interested in it.

[^1]: One of the key people involved in glibc's development history is Ulrich Drepper [who famously takes issues with static linking](https://akkadia.org/drepper/no_static_linking.html).

[^2]: People don't use Bedrock for its own stake, but to get access to other things.  Bedrock itself is essentially overhead we want to minimize.

[^3]: If you stick with stable Rust - meaning no build-std - glibc overhead is over a megabyte per binary.  My long-term low-priority target is to get Bedrock's core/essential bits in total to be under two megabytes.

[^4]: I'm a bit peeved Bedrock Linux has to fight with Minecraft server over SEO concerns, and [I'm sympathetic to others in similar situations](https://en.wikipedia.org/wiki/Go_(programming_language)#Naming_dispute).

[^5]: I have frustratingly little time to work on Bedrock, which results in most "development" occurring when my hands are busy but mind is free, e.g. grocery shopping, showering, lifting, etc.  By its very nature, this makes it hard to take notes.

[^6]: I consider it important to be able to fix a hypothetical high-priority bug while sleep deprived.

**Comment by Animeshz** [comment]
> I'm disinclined from making Naga's code public until it's ready for a public alpha <sup>[[5]](https://github.com/bedrocklinux/bedrocklinux-userland/issues/100#user-content-fn-5-9de7ee2ac88e3cc5fd5c20e38890392a)</sup>.

Why not? I think it'll be easier to develop different components in parts. People can understand the code in their free time and develop as with the plans (if posted), and may extend the plans under the discussion thread. I don't think much question will be asked, even if it does, it'll result more in a discussion where public opinions may be put and acted upon, also other community members can reply too :thinking:

**Comment by paradigm** [comment]
> > I'm disinclined from making Naga's code public until it's ready for a public alpha [[5]](https://github.com/bedrocklinux/bedrocklinux-userland/issues/100#user-content-fn-5-9de7ee2ac88e3cc5fd5c20e38890392a).
> 
> Why not?

Naga is taking a very long time not because there's so much work to do, but because the limited resources available to do it are consumed by low/no effort community interactions.  Publishing Naga's early source code would result in substantially more such things and stall it out completely.  It'll be faster in the long run for me to get it to a certain minimum level of polish before publishing anything.

> I think it'll be easier to develop different components in parts. People can understand the code in their free time and develop as with the plans (if posted), and may extend the plans under the discussion thread. I don't think much question will be asked, even if it does, it'll result more in a discussion where public opinions may be put and acted upon, also other community members can reply too in case you're busy thinking

Your thoughts do not match my experience.

Not to be rude, but I think it worth highlighting here that you've just asked an question that I answered in the very foot note link you quoted and consumed time that I would have otherwise spent working on Naga.  This very interaction is evidence in favor of my concern and against your conjecture.
